active/Cross Site WebSocket Hijacking.js

```diff
@@ -0,0 +1,100 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * 
+ * Author: Piotr Furman <piotr.furman at motorolasolutions.com>
+ */
+
+var Base64 = Java.type("java.util.Base64")
+var Random = Java.type("java.util.Random")
+var String = Java.type("java.lang.String")
+var ByteArray = Java.type("byte[]")
+
+var RISK = 3
+var CONFIDENCE = 2
+var TITLE = "Cross-Site WebSocket Hijacking"
+var DESCRIPTION = "Server accepted WebSocket connection through HTTP Upgrade request with modified Origin header."
+var SOLUTION = "Validate Origin header on WebSocket connection handshake, to ensure only specified origins are allowed to connect. Also, WebSocket handshake should use random tokens, similar to anti CSRF tokens."
+var REFERENCE = "https://tools.ietf.org/html/rfc6455#section-10.2"
+var OTHER = "See also https://portswigger.net/web-security/websockets/cross-site-websocket-hijacking or https://christian-schneider.net/CrossSiteWebSocketHijacking.html"
+var CWEID = 346 // CWE-346: Origin Validation Error, http://cwe.mitre.org/data/definitions/346.html
+var WASCID = 9 // WASC-9 Cross Site Request Forgery, http://projects.webappsec.org/w/page/13246919/Cross%20Site%20Request%20Forgery
+
+function scanNode(as, msg) {
+    var target = msg.getRequestHeader().getURI().toString()
+
+    // check if this is a WebSocket HTTP Upgrade request
+    // request message should include also "Connection: Upgrade" header but let's skip it for now, and switch to msg.isWebSocketUpgrade() if https://github.com/zaproxy/zaproxy/pull/6508 was merged
+    var upgrade_header = msg.getRequestHeader().getHeader("Upgrade")
+    if (!upgrade_header || upgrade_header.toLowerCase() !== "websocket") {
+        print("Cross-Site WebSocket Hijacking scan skipped for url=" + target + ", is not a WebSocket upgrade request")
+        return
+    }
+
+    print("Cross-Site WebSocket Hijacking scan started for url=" + target)
+    msg = msg.cloneRequest()
+    
+    // set random Sec-WebSocket-Key 
+    var randomBytes = new ByteArray(16)
+    new Random().nextBytes(randomBytes)
+    var secWsKey = new String(Base64.getEncoder().encode(randomBytes))
+    msg.getRequestHeader().setHeader("Sec-WebSocket-Key", secWsKey)
+
+    // set Origin header using custom domain, .example is a reserved TLD in RFC 2606 so it should not match domain name of a scanned service
+    msg.getRequestHeader().setHeader("Origin", "https://cswsh.example")
+
+    as.sendAndReceive(msg, true, false)
+
+    if (as.isStop()) {
```

Comment:
This can be removed, or if you think it's really needed do before sending the message. At this point it might as well do the quick status code check.

Reply:
Thanks, done.